### PR TITLE
Implement Serilog Activity enricher

### LINKED ISSUE #3453: [Serilog Extensions] Add Activity Enricher
We have a [Serilog Extensions](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Extensions.Serilog) project now which currently contains a sink for routing logs emitted through Serilog into an `OpenTelemetryLoggerProvider`.

One thing that might be nice to also have in that project is an [enricher](https://github.com/serilog/serilog/wiki/Enrichment) which will push Activity/span data onto logs.

The use case for this would be Serilog being used for logs into some other sink (flat files, DB, whatever) and OpenTelemetry being used for Traces/Metrics.

Something like this...

```csharp
ILogger logger = new LoggerConfiguration()
    .Enrich.WithOpenTelemetry() // <- Registers enricher
    .WriteTo.RollingFile(
        new JsonFormatter(renderMessage: true), 
        @"C:\logs\log-{Date}.txt")    
    .CreateLogger();
```

### PR DESCRIPTION
**Pull request by wilsonrivera**
Fixes #3453.

## Changes

Implement a Serilog enricher which adds the `SpanId`, `TraceId`, `ParentSpanId` and `ActivityTraceFlags` properties to log events when the `Activity.Current` is not null, otherwise, these properties are not added

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by wilsonrivera** [comment]
Oh no :/ I made the changes on my 'main' branch and didn't notice until now, I hope it doesn't matter

File: src/OpenTelemetry.Extensions.Serilog/OpenTelemetrySerilogEnricher.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+            Activity? activity = Activity.Current;
+            if (activity == null)
+            {
+                return;
+            }
+
+            logEvent!.AddPropertyIfAbsent(propertyFactory!.CreateProperty(nameof(Activity.SpanId), activity.SpanId.ToHexString()));
+            logEvent.AddPropertyIfAbsent(propertyFactory.CreateProperty(nameof(Activity.TraceId), activity.TraceId.ToHexString()));
+            logEvent.AddPropertyIfAbsent(propertyFactory.CreateProperty(nameof(Activity.ParentSpanId), activity.ParentSpanId.ToHexString()));
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions.Serilog/OpenTelemetrySerilogEnricher.cs:40]: ParentSpanId could be default.

Comment (wilsonrivera) [reply src/OpenTelemetry.Extensions.Serilog/OpenTelemetrySerilogEnricher.cs:40]: You are correct, should it be skipped when default?

Comment (cijothomas) [reply src/OpenTelemetry.Extensions.Serilog/OpenTelemetrySerilogEnricher.cs:40]: yes.

Comment (wilsonrivera) [reply src/OpenTelemetry.Extensions.Serilog/OpenTelemetrySerilogEnricher.cs]: I've corrected this, I don't know if you or I should mark the conversation as solved

**Comment by cijothomas** [comment]
@CodeBlanch I am now unsure if this is something OTel repo should provide? Its like a general serilog enricher, for enriching with `Activity.Current`. This does not have anything to do with sending serilog logs via OTel pipeline 🙌

**Comment by wilsonrivera** [comment]
> @CodeBlanch I am now unsure if this is something OTel repo should provide? Its like a general serilog enricher, for enriching with `Activity.Current`. This does not have anything to do with sending serilog logs via OTel pipeline 🙌

After a small Google search, I found out this https://github.com/RehanSaeed/Serilog.Enrichers.Span project which seems to do exactly just this, could be a better solution in the long run

**Comment by CodeBlanch** [comment]
@wilsonrivera We should probably update the README a bit to explain the different things offered, use cases, etc. Feel like taking that on?

@cijothomas There is a bit on #3453 about the use case. I think this would be good to have. Why? Just to be helpful 🤣 Once a package with "OpenTelemetry" & "Serilog" exists on NuGet people will just naturally look there for this type of thing and it being from the same repo yields some confidence it will work correctly with everything else. Just my opinion!

**Comment by wilsonrivera** [comment]
Yeah, I can do that!

**Comment by CodeBlanch** [comment]
@wilsonrivera An update... we discussed on the weekly .NET SIG call today this PR and we have decided that it is good to be included. Later on we will likely move this project to the "contrib" repo and we might give it a dedicated package at that time, or we might try to contribute to serilog itself, but we don't need to tackle that now. I think we just need to do README updates and then we can get this merged.

**Review by cijothomas — APPROVED** [review]

**Comment by pdonovan** [comment]
> @wilsonrivera An update... we discussed on the weekly .NET SIG call today this PR and we have decided that it is good to be included. Later on we will likely move this project to the "contrib" repo and we might give it a dedicated package at that time, or we might try to contribute to serilog itself, but we don't need to tackle that now. I think we just need to do README updates and then we can get this merged.

Not sure the best place to ask this... but when we should end-users (e.g. myself) expect to get a prerelease NuGet package for OpenTelemetry.Extensions.Serilog? I tried copying the code but it won't build because it uses internal classes. Our project is invested in Serilog, and logging associated with traces is the last thing we need in our OTEL evaluation.

**Comment by CodeBlanch** [comment]
@pdonovan I spun up an issue to track your question: #3602

**Comment by wilsonrivera** [comment]
Sorry for the delay, life got in the way but I've wrote a little documentation on how to setup the activity enricher, directly on the Serilog project

**Comment by pdonovan** [comment]
> Sorry for the delay, life got in the way but I've wrote a little documentation on how to setup the activity enricher, directly on the Serilog project

Link? I had a quick look around the [GitHub serilog pages](https://github.com/serilog/serilog) but couldn't find anything.

**Pull request closed without merging**